src/Microsoft.DotNet.XHarness.WebAssembly/xunit-runner.cs

```diff
@@ -0,0 +1,354 @@
+        private List<ITestCase> _testCases;
+        private readonly XunitProject _project;
+        private string _last_name = string.Empty;
+        public int nrun, nfail, nskipped, nfiltered;
+
+        IEnumerable<ITestCase> TestCases {
+            get {
+                if (_testCases == null)
+                {
+                    _testCases = new List<ITestCase>();
+                    _discoverer.Discover();
+                    Console.WriteLine($"Discovered tests: {_testCases.Count }");
+                }
+
+                foreach (ITestCase testCase in _testCases)
+                {
+                    yield return testCase;
+                }
+            }
+        }
+
+        public WasmRunner(XunitProject project)
+        {
+            _project = project;
+
+            string assemblyFileName = "/" + project.Assemblies.FirstOrDefault().AssemblyFilename;
+
+            var assembly = Assembly.LoadFrom(assemblyFileName);
+            var assemblyInfo = new ReflectionAssemblyInfo(assembly);
+
+            IAttributeInfo collectionBehaviorAttribute = assemblyInfo.GetCustomAttributes(typeof(CollectionBehaviorAttribute)).SingleOrDefault();
+            var testAssembly = new TestAssembly(assemblyInfo, null);
+
+            IXunitTestCollectionFactory collectionFactory = ExtensibilityPointFactory.GetXunitTestCollectionFactory(this, collectionBehaviorAttribute, testAssembly);
+
+            _discoveryOptions = TestFrameworkOptions.ForDiscovery(null);
+            _discoverer = new Discoverer(assemblyInfo, new NullSourceInformationProvider(), this, _discoveryOptions, collectionFactory);
+        }
+
+        static object ConvertArg(object arg, Type argType)
+        {
+            if (arg == null || arg.GetType() == argType)
+                return arg;
+            // Not clear what conversions should be done
+            // IEnumerable<T> -> T[]
+            if (argType.IsArray)
+            {
+                Type t = typeof(IEnumerable<>).MakeGenericType(argType.GetElementType());
+                if (t.IsAssignableFrom(arg.GetType()))
+                {
+                    MethodInfo m = typeof(Enumerable).GetMethod("ToArray").MakeGenericMethod(new Type[] { argType.GetElementType() });
+                    return m.Invoke(null, new object[] { arg });
+                }
+            }
+            try
+            {
+                if (arg != null && arg is IConvertible)
+                    arg = Convert.ChangeType(arg, argType);
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine("EX (ConvertArg): " + ex);
+            }
+
+            return arg;
+        }
+
+        public virtual bool OnMessage(IMessageSinkMessage msg)
+        {
+            if (msg is Xunit.Sdk.TestCaseDiscoveryMessage disc_msg)
+            {
+                _testCases.Add(disc_msg.TestCase);
+                return true;
+            }
+
+            return true;
```

Comment:
If we always return true, we should just have an exit point for the method, less complexity.

Reply:
Removed redundant return